src/py2spack/package_providers.py

```diff
@@ -0,0 +1,316 @@
+    ) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Get the contents of the pyproject.toml file for the specified version."""
+
+    @abc.abstractmethod
+    def get_hash(
+        self, name: str, version: vn.Version
+    ) -> dict[str, str] | PyProjectProviderQueryError:
+        """Get the sdist hash (sha256 if available) for the specified version."""
+
+
+class PyPIProvider(PyProjectProvider):
+    """Obtains project versions and distribution packages through the PyPI JSON API.
+
+    Various public and private methods of this class cache their return values in order
+    to minimize the number of requests to the provider.
+    """
+
+    def __init__(self, base_url: str = "https://pypi.org/simple/") -> None:
+        """Initialize PyPI Provider with base API url."""
+        if not base_url.endswith("/"):
+            base_url = f"{base_url}/"
+        self.base_url = base_url
+
+    @functools.lru_cache(maxsize=1)  # noqa: B019
+    def _get(self, name: str) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Load info for the available distribution files from PyPI.
+
+        Data for the most recent package is cached. The cache has a maxsize of 1,
+        because during conversion of a package, both methods `get_versions` as
+        well as `get_pyproject` use the returned data and we want to omit re-
+        peated requests to the API. Since all the method calls happen together,
+        before another package is requested, the cache size of 1 is enough.
+        """
+        r = requests.get(
+            f"{self.base_url}{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != utils.HTTP_STATUS_SUCCESS:
+            if r.status_code == utils.HTTP_STATUS_NOT_FOUND:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return PyProjectProviderQueryError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return PyProjectProviderQueryError(msg)
+
+        data: dict[Any, Any] = r.json()
+        return data
+
+    @functools.cache  # noqa: B019
+    def get_versions(self, name: str) -> list[vn.Version] | PyProjectProviderQueryError:
+        """Get usable versions for package.
+
+        Returns an error if no versions are found.
+        In addition to the caching of the `_get` method, we also cache all calls
+        to `get_versions`, because the versions are needed frequently during the
+        conversion process for dependencies, and the size of the data is small.
+        """
+        data = self._get(name)
+        if isinstance(data, PyProjectProviderQueryError):
+            return data
+
+        versions = data["versions"]
+
+        # parse and sort versions
+        result: list[vn.Version] | PyProjectProviderQueryError = sorted(
```

Comment:
It looks to me like this type annotation is wrong and only lists (or empty lists) can be produced here, am I wrong?
```suggestion
        result: list[vn.Version] = sorted(
```

Reply:
correct